mir_simulation/mir_world_generation/common/mir_world_generation/node.py

```diff
@@ -0,0 +1,109 @@
+import math
+import random
+
+class Node(object):
+
+    """Docstring for Node. """
+
+    ws_length = 80 # cm
+    ws_width = 50 # cm
+    def __init__(self, col=0, row=0, size=150):
+        self.size = size
+        self.x = col * self.size
+        self.y = row * self.size
+        self.ws = []
+        # North, South, East, West (north will be positive Y axis)
+        self.probable_ws_direction = ['N', 'S', 'E', 'W'] 
+
+    def add_ws(self):
+        min_len_pos = self.ws_length/2
+        max_len_pos = self.size - (self.ws_length/2)
+        min_wid_pos = self.ws_width/2
+        max_wid_pos = self.size - (self.ws_width/2)
+
+        direction = random.choice(self.probable_ws_direction)
+        self.probable_ws_direction.remove(direction)
+        if direction == 'N':
+            if 'S' in self.probable_ws_direction:
+                self.probable_ws_direction.remove('S')
+            x = self._resolve_overlap(direction) if len(self.ws) > 0 \
+                    else random.randint(self.x + min_len_pos, self.x + max_len_pos)
+            y = self.y + min_wid_pos
+            theta = math.pi/2
+        if direction == 'S':
+            if 'N' in self.probable_ws_direction:
+                self.probable_ws_direction.remove('N')
+            x = self._resolve_overlap(direction) if len(self.ws) > 0 \
+                    else random.randint(self.x + min_len_pos, self.x + max_len_pos)
+            y = self.y + max_wid_pos
+            theta = -math.pi/2
+        if direction == 'E':
+            if 'W' in self.probable_ws_direction:
+                self.probable_ws_direction.remove('W')
+            x = self.x + max_wid_pos
+            y = self._resolve_overlap(direction) if len(self.ws) > 0 \
+                    else random.randint(self.y + min_len_pos, self.y + max_len_pos)
+            theta = 0.0
+        if direction == 'W':
+            if 'E' in self.probable_ws_direction:
+                self.probable_ws_direction.remove('E')
+            x = self.x + min_wid_pos
+            y = self._resolve_overlap(direction) if len(self.ws) > 0 \
+                    else random.randint(self.y + min_len_pos, self.y + max_len_pos)
+            theta = math.pi
+
+        self.ws.append({'x': x, 'y': y, 'theta': theta})
+
+    def _resolve_overlap(self, direction):
```

Comment:
can you explain a little bit what's going on here?

Reply:
The generation part works with a grid of cells. Each cell is 1.5 x 1.5 meters. These dimensions allow it to place the robot and 2 workstation (ws, sh, pp) in a cell simultaneously with a bit more room for the robot to rotate in place and move back/forward if required (required for shelf pick).
 
When there is only one workstation in a cell, that workstation can be generated anywhere along a wall in a cell. However, when there are 2 workstation, they need to be placed in a certain specific way (they have to be perpendicular to each other and place next to diagonal corners) in order for them not to collide with each other and for the robot to be able to access both of them.

Now, since the workstation are spawned individually and randomly, the first workstation will be spawned randomply along a wall in a cell and if a second workstation needs to be in the cell, then, the first workstation needs to be moved to a corner and the second one needs to be spawned at the diagonally opposite corner. The function `_resolve_overlap` does exactly this task.

Reply:
ah ok, makes sense. Can that be compressed into a 2 line documentation for that function? :)

Reply:
Sure. I can add docstrings to that. I didn't add any earlier cause I generally don't add docstrings to private functions.